(TK-316) client metrics support

rlinehan opened the pull request:
This PR adds metrics support to the http client (clojure and java, sync
and async). A metric registry can optionally be passed into the client on
creation. If a metric registry is present, a metric will be added to time each
request. The metric name is currently `puppetlabs.http-client.<url>.<HTTP
method>` but in the future (for TK-317) there will be a metric-id request
option.

This PR also adds a `get-client-metrics`/`.getClientMetrics` function that
takes a client instance and returns the http client-specific metrics from the
metric registry and a `get-client-metrics-data`/`.getClientMetricsData` function that takes a client instance and returns a map of metrics information that can be added to a status endpoint.

rlinehan commented:
@scottyw (or anyone else looking at this) - I've found that this doesn't seem to work as expected (or the tests I'm trying to write are really wrong). Here's a commit illustrating what I've found: https://github.com/rlinehan/clj-http-client/commit/5bd0ab4223f0308177691c7c139693c8be2dc989

scottyw commented:
Theory here: https://github.com/rlinehan/clj-http-client/pull/1

rlinehan commented on src/clj/puppetlabs/http/client/async.clj:
```diff
@@ -119,9 +121,38 @@
+        aggregate (* count mean)]
+    {:count count
+     :mean mean
+     :aggregate aggregate}))
```
This is designed to match what is being returned by [trapperkeeper-comidi-metrics](https://github.com/puppetlabs/trapperkeeper-comidi-metrics/blob/6211f2621438cedc484ba0cd3c10d0d97a4a8106/src/puppetlabs/metrics/http.clj#L84-L98). It seems like it might be good to separate some of that out into a library (that doesn't have a trapperkeeper dependency) that could be used here.

rlinehan commented:
This PR still has a bug for metrics on requests that use unbuffered streaming, but I wanted to put this up as-is in case anyone had time to provide feedback on what is currently here.

scottyw commented on test/puppetlabs/http/client/async_unbuffered_test.clj:
```diff
@@ -318,7 +319,7 @@
 (deftest java-existing-streaming-with-small-payload-with-decompression
   (testing "java :stream with 1K payload and decompression"
-    (java-blocking-streaming (generate-data 1024) ResponseBodyType/STREAM false)))
+    (java-blocking-streaming (generate-data 1024) ResponseBodyType/STREAM true)))
```
Good spot!

rlinehan commented on test/puppetlabs/http/client/async_unbuffered_test.clj:
```diff
@@ -327,3 +328,167 @@
+                  (is (> (:mean unbuffered-stream-data) (:mean metric-data))))))))
+         (catch TimeoutException e
+           ;; Expected whenever a server-side failure is generated
+           ))))))
```
These tests are kind of gross and repetitive, but I didn't really want to overload the `java-non-blocking-streaming` and `clojure-non-blocking-streaming` functions even more, and I didn't want to start down a rabbit hole of refactoring this whole namespace.

scottyw commented on src/java/com/puppetlabs/http/client/impl/JavaClient.java:
```diff
@@ -324,27 +337,52 @@ public static void requestWithClient(final RequestOptions requestOptions,
         final HttpContext httpContext = HttpClientContext.create();
 
-        final FutureCallback<HttpResponse> futureCallback = new FutureCallback<HttpResponse>() {
+        final Timer.Context timerContext = registry == null ? null : timer(registry, request).time();
```
This need to start the timer outside the ```TimedFutureCallback``` but stop it inside seems slightly funny to me. I wonder would it be nice to make ```TimedFutureCallback``` a delegate that takes a registry, request and callback as constructor args? We could add a ```start()``` method too that could be called closer to the moment Apache's ```execute()``` method is called. That might remove the need for the separate ```timedXxxx()``` methods too.

scottyw replied:
I had a go to see what this looked like: https://github.com/rlinehan/clj-http-client/pull/2

rlinehan replied:
I think I like that. I think not needing the separate `timedXxx()` methods is a lot cleaner. I'll play with it a bit tomorrow. Thanks!

scottyw commented on src/java/com/puppetlabs/http/client/impl/JavaClient.java:
```diff
@@ -268,7 +278,7 @@ private static void executeWithConsumer(final CloseableHttpAsyncClient client,
             @Override
             public void deliver(HttpResponse httpResponse) {
-                futureCallback.completed(httpResponse);
+                futureCallback.completed(httpResponse); // this stops the timer for the request metric
```
FWIW I got confused here because this call also stops a timer context, but it's the one that was in scope when the ```TimedFutureCallback``` was created not the ```timerContext``` in scope here which is a different one.

rlinehan replied:
Yeah, it is confusing. Your commit makes it a bit easier to understand because it has more to do with what's in scope, but I still find it a bit confusing what timer is stopped when. I'll see if I can make that less confusing.

rlinehan commented:
@scottyw thanks for that commit and for your feedback! I made a couple small changes to it and merged it to this branch.

scottyw commented:
LGTM! This is such a great addition. Thank you.

cprice404 commented on src/clj/puppetlabs/http/client/async.clj:
```diff
@@ -119,9 +121,38 @@
+  (let [timers (get-client-metrics metric-registry)]
+    (into {} (map (fn [[metric-id timer]]
+                    {metric-id (assoc (get-metric-data timer) :metric-id metric-id)})
+                  timers))))
```
nit: someday I need to write a little benchmark for this, but it seems like `reduce` would be faster since it doesn't have to iterate over the list twice.  I also think it's more readable but apparently I'm the only person who does :)

rlinehan replied:
Addressed in 1488a45.

cprice404 commented on src/clj/puppetlabs/http/client/async.clj:
```diff
@@ -147,7 +178,8 @@
    callback :- common/ResponseCallbackFn
-   client :- HttpAsyncClient]
+   client :- HttpAsyncClient
+   metric-registry :- common/OptionalMetricRegistry]
```
this is technically an API change, right?

rlinehan replied:
Ah yeah. I was about to ask "what's our API", but I guess everything in the "Public" section we consider our API? It seems not super likely to be used given that we have `make-request` now, but it's very easy for me to make this not an API change, so I'll do that.

rlinehan replied:
Addressed in 58c1a3c.

cprice404 commented on src/java/com/puppetlabs/http/client/SyncHttpClient.java:
```diff
@@ -12,6 +15,16 @@
+    /**
+     * @return a Map of metric name to metric data
+     */
+    public Map<String, Map<String, Object>> getClientMetricsData();
```
Curious why we're using a SortedMap for one of these but not the other?

rlinehan replied:
`SortedMap` for the first because that was the return type of the underlying method from the dropwizard library. I can make both of these `Map` for consistency though.

rlinehan replied:
Address in 1488a45.

cprice404 commented on src/java/com/puppetlabs/http/client/impl/ClientMetricFilter.java:
```diff
@@ -0,0 +1,11 @@
+
+    public boolean matches (String name, Metric metric) {
+        return name.startsWith("puppetlabs.http-client");
+    }
```
I think this is fine but we should chat briefly about what happens if multiple Puppet Servers are all trying to export this data to graphite.  I *think* the graphite export code will give us a way to namespace / differentiate the metrics between servers, but we should double-check.

cprice404 commented on src/java/com/puppetlabs/http/client/impl/JavaClient.java:
```diff
@@ -493,4 +511,56 @@ public static Object coerceBodyType(InputStream body, ResponseBodyType as,
+    private static Timer.Context startTimer(MetricRegistry registry, HttpRequest request) {
+        if (registry != null) {
+            final RequestLine requestLine = request.getRequestLine();
+            final String name = MetricRegistry.name("puppetlabs.http-client", requestLine.getUri(),
```
Can we extract a constant for `puppetlabs.http-client`?

Also it looks like we're just using the raw URI string for the metrics names.  Did you have thoughts on how that maps to the 'metric-id' tuples that we discussed earlier?  I think that working the code up to that point incrementally is fine, but if we wanted to merge this chunk of work before those changes landed, I think we'd need to have a pretty clear idea of how we might evolve it to add those.  Perhaps it could be as simple as adding another namespace to the metric name in these code paths, e.g. `puppetlabs.http-client.unidentified-requests` or `raw-requests` or something.

rlinehan replied:
I wanted to get a PR up with all the metric plumbing and make sure that all was okay and then add in support for `metric-id` as a request option. The ticket for that is TK-317.

My thinking had been that `metric-id` would be optional, and if it wasn't specified use the raw URI and request method. However, I could see it maybe being useful to put all the requests without a metric-id under `unidentified-requests`?

rlinehan replied:
Constant added in 1488a45.

cprice404 commented on src/java/com/puppetlabs/http/client/impl/JavaClient.java:
```diff
@@ -493,4 +511,56 @@ public static Object coerceBodyType(InputStream body, ResponseBodyType as,
+        if (registry != null) {
+            final RequestLine requestLine = request.getRequestLine();
+            final String name = MetricRegistry.name("puppetlabs.http-client", requestLine.getUri(),
+                    requestLine.getMethod());
```
We will also want to be careful about the hierarchy for anything we include besides the URI/metric id.  Having the request method reflected in the metrics will be useful for sure, but we might want to think about putting it *before* the URI, in case we think of other sub-uri metrics we want to add on later.

Also, we will probably want to break these metrics into multiple timers so that we can request timing data at any level in the hierarchy; we can look at the metrics stuff in the 'PuppetProfiler' in pe-pse for examples and possibly some re-usable code.

rlinehan replied:
Hmm. What else are you thinking about having under the URI?

I was thinking that if you have a URI that can take multiple methods, you probably care more about grouping based on URI and then based on method than based on method first.

I think that having multiple timers going down the hierarchy probably makes sense. I guess one thought to counter the previous statement I made about request method is that if you had the ability to do both GET and POST at like /users/<user id>, you might want to get information about all GET/POST requests to just `/users/*`, and that would be easiest if we had timers along the hierarchy for `GET./users./<user-id>` and `POST./users./<user-id>` than for something like `/users./<user-id>.GET`.

cprice404 commented on src/java/com/puppetlabs/http/client/impl/JavaClient.java:
```diff
@@ -493,4 +511,56 @@ public static Object coerceBodyType(InputStream body, ResponseBodyType as,
+        }
+    }
+
+    public static SortedMap<String, Timer> getClientMetrics(MetricRegistry metricRegistry){
```
nit: it'd be nice to keep all of the public methods in one spot and the private methods in another spot, but we've kind of already broken that pattern before your PR, so, maybe not worth trying to fix everything.

Would be good to have  javadocs on the public methods, though.

rlinehan replied:
This namespace is under "impl" and is definitely a jumble of public and private methods, none of which have any javadocs (the javadocs are typically on the methods in the `(As|Sync)HttpClient` interfaces). I guess I can add javadocs to these new public methods if that's what we'd eventually like for all of the methods in this namespace?

cprice404 commented on src/java/com/puppetlabs/http/client/impl/JavaClient.java:
```diff
@@ -493,4 +511,56 @@ public static Object coerceBodyType(InputStream body, ResponseBodyType as,
+                data.put("count", count);
+                data.put("mean", meanMillis);
+                data.put("aggregate", count * meanMillis);
+                metricsData.put(metricId, data);
```
I'm a little waffly on this; it might be better to create a new class to hold the 4 fields and then put that class into a map instead of using `Object`.  Since we know that we're mostly going to be consuming this from Clojure I feel *slightly* less strongly about that, but I still think it might be preferable.

rlinehan replied:
Yeah, I did this because it seemed like this would be the easiest structure to consume from Clojure, but it does seem cleaner to have a new class for it.

rlinehan replied:
Addressed in a661205.

cprice404 commented on test/puppetlabs/http/client/async_unbuffered_test.clj:
```diff
@@ -327,3 +328,167 @@
+
+                    (is (> (get unbuffered-stream-data "mean") (get metric-data "mean")))))))))
+         (catch TimeoutException e
+           ;; Expected whenever a server-side failure is generated
```
What causes the server-side failure in this case?

Could we add a couple of pathological assertions `(is (true? false))` and `(is (true? true))` in the spots in the test code path that we do/don't expect it to reach due to the exception?

rlinehan replied:
So in my comment about these tests being gross and repetitive, what I probably should have mentioned is that these tests are mostly wholly copied from other tests in this namespace, and I don't 100% understand what is going on here. I can try to make at least these tests a little more readable though... I'm not sure I want to start down the rabbit whole of refactoring all the tests in the namespace.

cprice404 commented:
This is looking very good; I had a handful of nitpicks, but mostly I think the only real remaining work is to sort out how we want to organize the metrics names, support metric ids, etc., and it might be possible for us to break some of that into a separate PR.

rlinehan commented:
@cprice404 I've put up a few PRs addressing most of your feedback. Not addressed are some of the open questions about what timers we want and what we want the metric ids to look like, putting javadocs on the public methods in `JavaClient`, and clarifying the `catch` in the unbuffered stream failure tests.

rlinehan commented:
@cprice404 @scottyw I've pushed up some more commits implementing (I think) everything that's discussed in TK-316 and TK-317.

rlinehan commented on test/com/puppetlabs/http/client/impl/metrics_unit_test.clj:
```diff
@@ -0,0 +1,150 @@
+(ns com.puppetlabs.http.client.impl.metrics-unit-test
```
I'm not sure that this is the right place for these tests, because there are some tests of the clojure code here in addition to the java classes. I'm also not sure how much that matters.

camlow325 commented on src/clj/puppetlabs/http/client/sync.clj:
```diff
@@ -23,43 +23,63 @@
   [req]
   (with-open [client (async/create-default-client (extract-client-opts req))]
-    (request-with-client (extract-request-opts req) client)))
+    (request-with-client (extract-request-opts req) client nil)))
```
Seems like it would be good to have a way to pass in a `MetricsRegistry` even for one-off requests, not just for cases where you construct a full client?  To do that, we might need to have the client pass in the `MetricsRegistry` via an option in the req map?

rlinehan replied:
I guess this might be a good/useful thing. But I wonder, if someone is just making one-off requests, how much of a use-case is there for tracking metrics on them? If a user really wanted to be tracking metrics on a request, is it that hard for them to create a client and start using that? I don't have a super strong opinion about this, just pondering.

cprice404 replied:
Doesn't seem worth the ROI to me, personally.

rlinehan replied:
This is not specifically in response to the question of whether we should track metrics for one-off requests, but that is what got me thinking about it. The more I think about it, the more I wonder whether it would make a lot more sense to have the `MetricRegistry` passed in as a client option, rather than as its own argument to `create-client`. No other client option gets is own argument, and I think the only reason that I wrote the code this way to begin with is that I didn't quite understand how the client options worked.

cprice404 replied:
I'm +1 on the idea of passing it in as a client option.

camlow325 commented on src/java/com/puppetlabs/http/client/impl/ClientMetricFilter.java:
```diff
@@ -0,0 +1,64 @@
+import com.codahale.metrics.Metric;
+import org.apache.commons.lang3.StringUtils;
+
+public class ClientMetricFilter {
```
Just curious why you implemented the sub-filters as static classes under a `ClientMetricFilter` class.  Since it doesn't look like the subfilter classes share anything from `ClientMetricFilter`, I think it might look better to have a `metricfilter` namespace under `com.puppetlabs.http.client.impl` and just have these sub-filter classes be standalone classes that live in their own .java source files.  Would that make sense to you?

rlinehan replied:
Mostly implemented this way because I don't know how to java, and initially when implementing them I had thought that there would end up being even more if we ended up adding the `response-init` timers, and that they would all be only a couple lines, and it seemed silly to have 8 files with two lines. However, I think with the way the implementation worked out that it makes more sense to have these all be standalone classes (especially if that's the more java-y way of doing it).

camlow325 replied:
One other option might be to just create a single `MetricsFilter` class, `MetricFilterMatchingName` or something like that.  The constructor could take in a `String` object and use that object to do the `equals` comparison in a `matches` method callback.  Could then move the logic to construct the appropriate sub-namespace arrays back into calling `getClientMetrics*` methods in `Metrics` class?

camlow325 commented on src/java/com/puppetlabs/http/client/impl/PersistentSyncHttpClient.java:
```diff
@@ -26,10 +31,42 @@ private static void logAndRethrow(String msg, Throwable t) {
+        return Metrics.getClientMetrics(metricRegistry);
+    }
+
+    public Map<String, Timer> getClientMetrics(String url, Metrics.MetricType metricType) {
```
I wonder if it's worth it to repeat these `get*Metrics` helper methods across both of the HTTP client classes or just have clients interact with the underlying `Metrics` class directly.  I'd probably opt for the latter - better separation of concerns, I think.  If we do want these methods on both client classes, though, I'd be inclined to create an abstract base class which implements them and then have the two Persistent*HttpClient classes extend from that base class - so that we don't have to duplicate the implementation across the classes.

rlinehan replied:
I think at one point @cprice404 and I discussed this and decided that we leaned towards having these methods take the client (rather than require the metric registry to be passed in) because consumers of this already have to be holding on to a reference to the client. That said, I don't feel strongly about this, and if we do make it so that basic sync requests that don't use a client can have metrics tracked on them then we definitely need to change this. It would make testing easier too.

cprice404 replied:
I'm not too fussed about this either way, but it did seem easier to me for the users to not need to worry about holding references to both the `MetricsRegistry` and the client object, and correlating the two of them.  Seems easier to me for them to just grab a temporary local reference to the metrics registry from tk-metrics at the time when they're constructing the client, pass it into the constructor, and then not have to worry about it anymore afterward.

rlinehan replied:
One thing that is a little odd about having the "get"metrics APIs connected to a client object is that if you have one MetricRegistry in use by two client instances, calling getClientMetrics* on one of these clients will return metrics for requests from both clients. Looking at it from this point of view, I think this actually might be really confusing the way this is implemented and would be less confusing if the "get" metrics APIs operated on the MetricRegistry object instead.

camlow325 commented on src/java/com/puppetlabs/http/client/Sync.java:
```diff
@@ -86,8 +87,17 @@ private static Response request(SimpleRequestOptions simpleRequestOptions,
+     * @param metricRegistry a metric registry to track metrics on client requests
+     * @return A persistent synchronous HTTP client
+     */
+    public static SyncHttpClient createClient(ClientOptions clientOptions, MetricRegistry metricRegistry) {
```
Same comment here as with the Clojure side implementation.  I think it would be good to have a way to pass in a `MetricRegistry` even for one-off Sync requests like Sync.get().  In this case, we could provide that via another one of the `SimpleRequestOptions` or overload the existing methods with a new version that takes a `MetricRegistry` object.

camlow325 commented on src/clj/puppetlabs/http/client/async.clj:
```diff
@@ -117,11 +124,68 @@
 
+(schema/defn ^:always-validate get-client-metrics :- (schema/maybe common/Metrics)
+  "Returns the http client-specific metrics from the metric registry."
+  ([metric-registry :- common/OptionalMetricRegistry]
```
When would someone want to call this function without supplying a metric-registry?

rlinehan replied:
Since this is used for the implementation of the protocol function `get-client-metrics`, this could be called on a client that didn't have a metrics registry registered. Maybe this should be moved out of the "public" section? I'm not really clear on what exactly should go where.

cprice404 replied:
Yeah, this seems like it could be private.

camlow325 commented on src/clj/puppetlabs/http/client/async.clj:
```diff
@@ -117,11 +124,68 @@
+                                       metric-registry
+                                       (:url metric-filter)
+                                       (:verb metric-filter)
+                                       Metrics$MetricType/BYTES_READ))
```
When we get to the point where we are supporting more than one kind of metric, presumably we'll need to have some function to coerce the value from the metric-filter into the corresponding Java enum?

rlinehan replied:
Yep. I guess for the now I could have a function that is just hardcoded to return this, same as in the Java implementation.

camlow325 commented on src/clj/puppetlabs/http/client/common.clj:
```diff
@@ -149,3 +155,28 @@
+(def MetricFilter
+  (schema/conditional
+   #(contains? % :url)
+   {:metric-type (schema/enum "bytes-read")
```
May be worth putting this enum into its own def that both conditionals reference - so there's only one thing to update later when we add other metric-types?

cprice404 replied:
+1 on that.  Ideally I'd like to figure out a way to get rid of the `contains?` stuff in these, as it feels fragile.  That said, as long as it's not public API, I probably wouldn't block the PR on it.

cprice404 replied:
Also for the clojure API, I think the metric-type enums should contain keyword values rather than string values.

camlow325 commented on src/java/com/puppetlabs/http/client/impl/JavaClient.java:
```diff
@@ -57,8 +58,7 @@
 import java.security.cert.X509Certificate;
-import java.util.HashMap;
-import java.util.Map;
+import java.util.*;
```
Probably good to expand all of the individual imports used here rather than '*'.

rlinehan replied:
ARGH Intellij-- yes, agreed. Is there a setting for that (to tell Intellij not to do that when automatically importing things)?

camlow325 replied:
Yeah, I think I had to do something like this:

Editor -> Code Style -> Java -> Imports ->
  Class count to use import with '*' -> 9999
  Names count to use static import with '*': 9999

Not great but seems to work.

camlow325 commented on src/java/com/puppetlabs/http/client/impl/ClientMetricFilter.java:
```diff
@@ -0,0 +1,64 @@
+        public boolean matches(String s, Metric metric) {
+            String[] metricNameArray = { Metrics.METRIC_NAMESPACE, Metrics.URL_NAMESPACE, url,
+                    Metrics.metricTypeString(metricType) };
+            return s.equals(StringUtils.join(metricNameArray, "."));
```
For the places we join the raw names from the array to do the match, I wonder if it would be better to use the `MetricRegistry/name` helper method, like we do in `Metrics.startBytesReadUrlTimers`.  Seems like it would be best to use the same method everywhere that we join the segments of a name together?

camlow325 commented on src/java/com/puppetlabs/http/client/impl/Metrics.java:
```diff
@@ -0,0 +1,158 @@
+            for (int j = 0; j <= i; j++) {
+                currentId.add(metricId[j]);
+            }
+            currentId.add(0, ID_NAMESPACE);
```
If you put this line before the preceding for loop, could this just be `currentId.add(ID_NAMESPACE)`?

camlow325 commented on src/java/com/puppetlabs/http/client/impl/Metrics.java:
```diff
@@ -0,0 +1,158 @@
+                currentId.add(metricId[j]);
+            }
+            currentId.add(0, ID_NAMESPACE);
+            currentId.add(currentId.size(), BYTES_READ_STRING);
```
Could this just be `currentId.add(BYTES_READ_STRING)`?

camlow325 commented on src/java/com/puppetlabs/http/client/impl/Metrics.java:
```diff
@@ -0,0 +1,158 @@
+    public static Map<String, ClientMetricData> computeClientMetricsData(Map<String, Timer> timers){
+        Map<String, ClientMetricData> metricsData = new HashMap<>();
+        if (timers != null) {
+            for (SortedMap.Entry<String, Timer> entry : timers.entrySet()) {
```
Should this just be a `Map.Entry` rather than the more specific `SortedMap.Entry`?

camlow325 commented on src/java/com/puppetlabs/http/client/impl/TimedFutureCallback.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    public void completed(T result) {
+        if (timerContexts != null) {
```
Could put these if blocks in a common helper method that each of the callbacks invoke.

camlow325 commented on test/com/puppetlabs/http/client/impl/metrics_unit_test.clj:
```diff
@@ -0,0 +1,150 @@
+(defn add-metric-ns [string]
+  (str "puppetlabs.http-client.experimental." string))
+
+(deftest start-bytes-read-timers-test
```
Should there be `is` statements around each of the `(= ...)` blocks?  Seems like these tests would just be a no-op as currently written?

camlow325 commented on src/clj/puppetlabs/http/client/async.clj:
```diff
@@ -117,11 +124,68 @@
+    metric-filter :- common/MetricFilter]
+   (when metric-registry
+     (cond
+       (:verb metric-filter) (into {} (Metrics/getClientMetricsWithUrlAndVerb
```
nit - I think I'd prefer to refer to this as a "method" rather than a "verb".  Not a big deal, though.

camlow325 commented on test/com/puppetlabs/http/client/impl/metrics_unit_test.clj:
```diff
@@ -0,0 +1,150 @@
+                registry {:url "http://test.com" :verb "PUT" :metric-type "bytes-read"})))))
+    (testing "getClientMetricsData with metric id returns the right thing"
+      (let [java-data (Metrics/getClientMetricsDataWithMetricId
+                  registry (into-array ["foo"]) bytes-read)
```
Nit - the word `registry` on this - and a few of the lines in other assertions below - seems misaligned per the line above it.

camlow325 commented on test/puppetlabs/http/client/metrics_test.clj:
```diff
@@ -0,0 +1,481 @@
+            [puppetlabs.http.client.sync :as sync]
+            [puppetlabs.http.client.common :as common]
+            [puppetlabs.trapperkeeper.core :as tk])
+  (:import (com.puppetlabs.http.client.impl Metrics$MetricType ClientMetricData)
```
Nit - looks like the `Metrics$MetricType` import is unused.

camlow325 commented on test/puppetlabs/http/client/metrics_test.clj:
```diff
@@ -0,0 +1,481 @@
+                    (is (= bytes-read-id (.getMetricId bytes-read-data)))
+                    (is (<= 1 (.getMean bytes-read-data)))
+                    (is (<= 1 (.getAggregate bytes-read-data)))))))))
+         (catch TimeoutException e
```
Under what circumstances would it be normal for this exception to be thrown for this test?

rlinehan replied:
These tests are wrapped in `try/catch`'s because the webserver handling the client requests is being shutdown before it has completed handling these requests. Jetty throws a `TimeoutException` if its `stopTimeout` is reached before it gracefully completes all open requests. I've got a branch upgrading to tk-w-j9 1.5.4 (where we changed this behavior for TK-338) and removing these `try/catch`'s, I just wasn't sure it was a great idea to add that upgrade to a PR that's been open for a while and has a lot of commits.

camlow325 replied:
Cool, that makes sense.  My biggest concern had been about the `TimeoutException` potentially being thrown in some other cases, causing the bulk of the test to not be executed.

It does seem like there would be a way to avoid Jetty encountering an error during shutdown.  Like maybe we could inject a promise into the `blocking-handler` and have its `future` block on that promise after doing the stream write.  Once we get past having made all of the assertions from within this test, could deliver on that promise.  The `future` could then be unblocked and close its output stream.  If all this happens before the webserver is shut down, I'd hope the server could shut down cleanly - without timing out - and that we wouldn't have to artificially try to work around that anymore be setting `shutdown-timeout-seconds` to 1 in the web server configuration.

That said, I think you're just following the pattern of pre-existing unbuffered stream tests here and would be fine with deferring any broader cleanup around that set of tests to sometime later than this PR.

camlow325 commented on test/puppetlabs/http/client/metrics_test.clj:
```diff
@@ -0,0 +1,481 @@
+
+                    (is (= 1 (.getCount bytes-read-data)))
+                    (is (= bytes-read-id (.getMetricId bytes-read-data)))
+                    (is (<= 1 (.getMean bytes-read-data)))
```
Should we make the minimum higher than 1 (like 200, the configured socket timeout) for this and the next assertion?  1 seems pretty short for the configured client timeouts.

camlow325 commented on test/puppetlabs/http/client/metrics_test.clj:
```diff
@@ -0,0 +1,481 @@
+                (let [bytes-read-data (get client-metrics-data bytes-read-id)]
+                  (is (= {:count 1 :metric-id bytes-read-id}
+                         (select-keys bytes-read-data [:metric-id :count])))
+                  (is (<= 1 (:mean bytes-read-data)))
```
Same comment here about the minimum timeouts for this one as for the corresponding Java test.

camlow325 commented on test/puppetlabs/http/client/metrics_test.clj:
```diff
@@ -0,0 +1,481 @@
+                  (is (<= 1 (:aggregate bytes-read-data))))))))
+         (catch TimeoutException e
+           ;; Expected whenever a server-side failure is generated
+           ))))))
```
Same question here as for the corresponding Java test on when we'd expect it to be normal for this exception to be thrown.

camlow325 commented:
Done reviewing for now.  Things are looking really good to me so far.  I thought the tests were pretty comprehensive - nice job!

I left a few comments throughout.  Probably the most important ones to consider:

- Should we have a way to measure metrics for one-off requests (and not just for cases where a full-blown "client" object is established)?

- A few questions about the organization of the Java classes - e.g.:

-- Would it make sense to have the "get" metrics APIs be standalone from instances of a "client" object?  Could reduce some of the duplication and might make sense in particular if we want to support measuring metrics for one-off requests (where the "client" objects are never exposed outside of the library).

-- Maybe rework the Filter classes a bit - reducing to a single one that's parameterized by the matching string or as classes in separate namespaces?

- Should the various external APIs tolerate passing in a null registry and returning null or essentially just "require" that a registry be passed in and throw an exception if one is not passed in?

cprice404 commented on src/clj/puppetlabs/http/client/async.clj:
```diff
@@ -117,11 +124,68 @@
+  "Returns a map of metric-id to metric data summary."
+  ([metric-registry :- common/OptionalMetricRegistry]
+   (let [timers (get-client-metrics metric-registry)]
+     (get-metrics-data timers)))
```
I notice that these fns return an empty map in the case where the client was created without a MetricsRegistry.

It might be worth considering whether they should instead return `nil`, or even throw an exception.  Not 100% sure what I think is the best out of those three options, just food for thought.

cprice404 commented on src/clj/puppetlabs/http/client/async.clj:
```diff
@@ -145,20 +209,26 @@
        are `:text` and `:stream`, which will return a `String` or an
        `InputStream`, respectively.  Defaults to `:stream`.
    * :query-params - used to set the query parameters of an http request"
```
Sad as it is, this is pretty much the best documentation we have about the request options at the moment.  Probably worth adding something here about how to specify a metric id.

cprice404 commented on src/clj/puppetlabs/http/client/common.clj:
```diff
@@ -70,7 +74,8 @@
    (ok :as)               BodyType
-   (ok :query-params)     {schema/Str schema/Str}})
+   (ok :query-params)     {schema/Str schema/Str}
+   (ok :metric-id)        [schema/Str]})
```
I think it would be nice if the Clojure API tolerated keywords here in addition to Strings.  The Java API doesn't need to.

cprice404 commented:
Just had time to start poking at this in the REPL.  It's freaking awesome.  A few improvements I think we can make to the API, but this is going to be so useful!  Nice job.  Will poke more tomorrow hopefully.

cprice404 commented on src/clj/puppetlabs/http/client/async.clj:
```diff
@@ -117,11 +124,68 @@
+   (let [timers (get-client-metrics metric-registry)]
+     (get-metrics-data timers)))
+  ([metric-registry :- common/OptionalMetricRegistry
+    metric-filter :- common/MetricFilter]
```
I wasn't sure exactly the best place to put some general API feedback, but I hate how github won't let you "reply" to a comment that isn't attached to a line of code, and I think most of this comes back to the schema for the `MetricFilter`, so this seemed like a reasonable place.  Rolling up a few things here:

1. I think we may want to monkey with some of the names, e.g. `with-metric-id`, `bytes-read`, etc., but I feel like it should be pretty easy to put that off as like the last thing we do in this PR (or maybe even a follow-on PR).  So I'd vote for maybe belaying that conversation.
2. Constructing one of these `MetricFilter` schema objects by hand feels a bit cumbersome (only talking about Clojure here, I don't necessarily care whether the Java API is a bit cumbersome, because, you know, it's Java).  I think it might be worth creating some helper functions to help build these up.  Maybe, at the least, a function that you pass in a partial map and it fills in the rest with some default values?  Or, to get a bit more granular, maybe something like `(metric-id-filter ["puppetlabs"])` and `(url-filter "https://tickets.puppetlabs.com")`.
3. I notice that if I make a request to `https://confluence.puppetlabs.com` and then another one to `https://confluence.puppetlabs.com/` (with a trailing slash), these show up as two separate metrics on the `with-url` side of the house.  I think that's probably fine, and that if people wanted to aggregate those somehow they can always use a `metric-id`.  Just thought it was worth mentioning; it also made me wonder if we needed to think at all about special characters that might show up in a URL and whether there are any possible issues with how those might get mapped to a metric id.
4. Assuming we allow raw URLs (including funky chars), to show up in the `with-url` metric ids, we should probably think a little about what might happen if someone exports these metrics to graphite.  I've found in the past that because of general issues around character escaping, it can be challenging to use a GUI like grafana to render a metric if the metric id has special characters in it, because in grafana you might have to do something like prepend the characters with backslashes or URL-encode them in order for it to find them in graphite.  I *think* it's probably OK for us to say that people should only consider exporting the `with-metric-id` metrics to graphite, and that that would probably alleviate this issue, but just something worth thinking about and potentially codifying in some docs or something.
5. I notice that query strings are included in the `with-url` metric ids.  Mixed feelings on this; I can see how they might be useful, but I can also imagine that leading to a large proliferation of timers, and make it hard or impossible to aggregate the URL metrics for a given URL ignoring the differences in the query string.  So, I could imagine us making the call that we just truncate query strings from the `with-url` metric ids, or I could imagine us maybe saying that in a future update we might add another type of `MetricFilter` that supports a regex value for the `:url` field.  Not sure.
6. I think it would be useful to have a `MetricFilter` that allowed me to ask for all of the `with-metric-id` metrics, without having to also specify a specific `metric-id`.
7. When I make two requests, one with a metric-id of ["puppetlabs" "jira"] and one with a metric-id of ["puppetlabs" "confluence"], and then I call `get-client-metrics-data` with a filter that includes `:metric-id ["puppetlabs"]`, I *only* get back the aggregate data.  I think I was kind of expecting that I'd get that *and* the values for the child nodes in that hierarchy.  I can imagine both use cases being desirable, so maybe we could add something to the `MetricFilter` that would allow you to flag whether you wanted only the exact match or everything underneath that in the tree.  It's possible that we already discussed this in our design meeting and agreed on the current approach, I can't remember.

Sorry, I know that's a lot of stuff :)  And I realize that the requirements for this have grown a bunch as we've been hacking away at it, so thanks for your patience.

I don't necessarily think that everything above needs to be addressed for this PR, just wanted to capture my thoughts.  It might be worth us setting up another design huddle meeting to prioritize these questions, decide which things we think are important, in what order, etc.  And figure out which of them would be breaking API changes if we added them at a later date.  I'd like to get to a point where we have a plan for the minimum set of changes we think are necessary to get this merged, and it's OK if we don't think it's quite in a releaseable state after the first PR is merged, but I think we need to try to minimize the amount of time that the repo is not in a releaseable state, so deciding on what exactly we think are the minimum requirements for an initial API and what things could be added later without breaking that API seems like the way to go.

rlinehan replied:
>1. I think we may want to monkey with some of the names, e.g. `with-metric-id`, `bytes-read`, etc., but I feel like it should be pretty easy to put that off as like the last thing we do in this PR (or maybe even a follow-on PR).  So I'd vote for maybe belaying that conversation.

Yeah, that's all easy to change and I figured we'd probably want to change it over time.

>2. Constructing one of these `MetricFilter` schema objects by hand feels a bit cumbersome (only talking about Clojure here, I don't necessarily care whether the Java API is a bit cumbersome, because, you know, it's Java).  I think it might be worth creating some helper functions to help build these up.  Maybe, at the least, a function that you pass in a partial map and it fills in the rest with some default values?  Or, to get a bit more granular, maybe something like `(metric-id-filter ["puppetlabs"])` and `(url-filter "https://tickets.puppetlabs.com")`.

Sure? I guess I don't feel like it's that cumbersome/makes that much of a difference - `{:metric-id [:foo] :metric-type :bytes-read}` vs `(metric-id-filter [:foo] :bytes-read)`, `{:url "http://tickets.puppetlabs.com :metric-type :bytes-read}` vs `(url-filter "https://tickets.puppetlabs.com" :bytes-read)`, and` {:url "http://tickets.puppetlabs.com :method "GET" :metric-type :bytes-read}` vs `(url-fitler "https://tickets.puppetlabs.com" "GET" :bytes-read)`? Or we have `(bytes-read-filter ...)` and `(response-init-filter ...)` that handle adding in the metric-type? This seems like something that can be added in later without being a breaking change, so long as we keep the schema the same.

One thing I realize from this though is that maybe we should standardize the method name schema should be keywords :get, :post, etc, and then we convert that to what the method will show up as in the metric name.

>3. I notice that if I make a request to `https://confluence.puppetlabs.com` and then another one to `https://confluence.puppetlabs.com/` (with a trailing slash), these show up as two separate metrics on the `with-url` side of the house.  I think that's probably fine, and that if people wanted to aggregate those somehow they can always use a `metric-id`.  Just thought it was worth mentioning; it also made me wonder if we needed to think at all about special characters that might show up in a URL and whether there are any possible issues with how those might get mapped to a metric id.

It seems maybe reasonable to strip trailing slashes... but then do we also strip trailing question marks? What else should we strip? What about things that can optionally percent encoded... should we standardize those before making the metric id? Lollll terrible questions. Okay, I think based on going down this train of thought I am against standardizing the url in any way before making it a part of a metric name - that's why we have the metric id feature.

>5. I notice that query strings are included in the `with-url` metric ids.  Mixed feelings on this; I can see how they might be useful, but I can also imagine that leading to a large proliferation of timers, and make it hard or impossible to aggregate the URL metrics for a given URL ignoring the differences in the query string.  So, I could imagine us making the call that we just truncate query strings from the `with-url` metric ids, or I could imagine us maybe saying that in a future update we might add another type of `MetricFilter` that supports a regex value for the `:url` field.  Not sure.

I actually didn't realize that query strings were included in this. I think that they should be stripped out - I agree that they'll make for too many too specific timers, making it hard to aggregate.

>6. I think it would be useful to have a `MetricFilter` that allowed me to ask for all of the `with-metric-id` metrics, without having to also specify a specific `metric-id`.

Sure.


>7. When I make two requests, one with a metric-id of ["puppetlabs" "jira"] and one with a metric-id of ["puppetlabs" "confluence"], and then I call `get-client-metrics-data` with a filter that includes `:metric-id ["puppetlabs"]`, I *only* get back the aggregate data.  I think I was kind of expecting that I'd get that *and* the values for the child nodes in that hierarchy.  I can imagine both use cases being desirable, so maybe we could add something to the `MetricFilter` that would allow you to flag whether you wanted only the exact match or everything underneath that in the tree.  It's possible that we already discussed this in our design meeting and agreed on the current approach, I can't remember.

I think I had asked about this in hipchat, and I think in that discussion we had decided to do just the exact match. But I think a reasonable API would be something like ["puppetlabs" "*"] if you want puppetlabs and everything under it, and ["puppetlabs"] if you just want the puppetlabs metric

cprice404 replied:
> Sure? I guess I don't feel like it's that cumbersome/makes that much of a difference - {:metric-id [:foo] :metric-type :bytes-read} vs (metric-id-filter [:foo] :bytes-read), {:url "http://tickets.puppetlabs.com :metric-type :bytes-read} vs (url-filter "https://tickets.puppetlabs.com" :bytes-read), and{:url "http://tickets.puppetlabs.com :method "GET" :metric-type :bytes-read} vs (url-fitler "https://tickets.puppetlabs.com" "GET" :bytes-read)? Or we have (bytes-read-filter ...) and (response-init-filter ...) that handle adding in the metric-type? This seems like something that can be added in later without being a breaking change, so long as we keep the schema the same.

What I was getting at is that I think we should have a default value that we use for the `:bytes-read` part, because I don't think users are probably going to want to have to care about that in the majority of cases.  So the value in what I'm suggesting would be that we'd be giving them a way to build a filter without having to express that part.  It could be an optional argument to one of these functions, but in most cases hopefully they wouldn't need it.

I also think that that schema is pretty likely to evolve over time, so if we can steer things in a direction where the "public API" is to call a filter-creation function, e.g. `(metrid-id-filter ...)`, then we're only committed to maintaining the integrity of the function signature(s) as API, and we can move the schema around as much as we need in the future.

> Okay, I think based on going down this train of thought I am against standardizing the url in any way before making it a part of a metric name - that's why we have the metric id feature.

I think I'm good with that, just wanted to think it through, and consider codifying it in docs or something.

> I actually didn't realize that query strings were included in this. I think that they should be stripped out - I agree that they'll make for too many too specific timers, making it hard to aggregate.

I think that sounds best to me too.  I still think that regex support on the URL filter would be valuable but that definitely seems like something that could be additive API in the future.  Might be worth capturing in Jira if we all agree that it would be useful, though.

> I think I had asked about this in hipchat, and I think in that discussion we had decided to do just the exact match. But I think a reasonable API would be something like ["puppetlabs" "*"] if you want puppetlabs and everything under it, and ["puppetlabs"] if you just want the puppetlabs metric

Yeah, that may very well be the case.  My memory is terrible.  :)  This also seems like something we could probably add in the future without breaking API, right?  If so then maybe a Jira for that is sufficient for now.

camlow325 replied:
:+1: on stripping the query param from the URL metrics.  I also didn't clue into the fact that it was being included.

rlinehan commented:
@cprice404 @camlow325 I believe I have addressed all of your feedback.

cprice404 commented:
@rlinehan nice job, thanks.  I just finished another review at the API level and wanted to try to capture my thoughts.  Would like to do one more review at the code level once we're all happy on the API front, but I'll wait to hear your thoughts on these comments.

I'm trying to break up these comments/suggestions into three tiers: a) things that I think have a significant enough impact on the future of the API that we may want to get them sorted before this PR is merged; b) things that we can break out into subsequent PRs but probably ought to try to get in before we do an official release, and c) stuff we can tackle opportunistically in the future.

For a) - things I think are at least worth discussing for this PR (and I could be convinced to punt these to another PR if others think that is the way to go): 
1. This is my fault, you just followed my suggestion from my previous comment about how to name the filter-creation helper functions (`metric-id-filter` etc.)... so, apologies.  But, when poking around with those functions in the IDE, I found myself hoping that the function names would all have a common prefix so that I could start typing `filter-` and then the editor would be able to show me a list of autocompletion choices, and from that I'd know exactly what my options were, as opposed to having to go find where those things live together in the code or look at docs, etc.  So, how would you feel about renaming those to something like `filter-metric-id` of `filter-by-metric-id` or `filter-with-metric-id` or something along those lines?
2. If I just make a call to `get-client-metrics`, the results that come back are pretty homogenous even though the data being returned is decidedly not.  I wonder if we should do some combination of the following: a) make the return value from that function have a top-level map with two keys in it, like `:metric-ids` and `:urls`, or some variant of that?  and then the individual metrics would live in nested maps underneath those?  b) should we be putting the metric-id tuple into the results as another key/value pair in those maps?  c) should we be using the metric-id tuple as the key for the metrics that have a metric id?  I'm just trying to imagine use cases where something like a status endpoint, which may not have any visibility into the places in code where the http requests are actually being made, wants to display a top-10 list of metrics that have a metric id, and use the metric id as part of the visualization; and then possibly a separate top-10 list of metrics by URL?

For b) - things I think we might want to do before release but am fine with not including in this PR; maybe they should be ticketed:
1. Adding a filter fn for "all metrics with any metrics id" (this might be unnecessary if we reorganize the data structure returned from `get-client-metrics` so that there was just a nested map in there that had these metrics in it).
2. Talk about what we want the final names to be for `with-*` and `bytes-read`.
3. Maybe play with graphite and validate that it's possible to access metrics that have funky characters in them, e.g. from a URL being embedded in a metric name

For c) - stuff that we should or could do but doesn't need to block the initial release per se:
1. Docs - and this doesn't have to be all @rlinehan , we already had a docs deficiency in this repo, it just seems like it's even more important now
2. Look into cleaning up schema for MetricFilter
3. Ability to filter URL metrics via regex

@rlinehan let me know what you think about all of that stuff.  Happy to dismiss or reschedule parts of that if you disagree with any of them.

rlinehan commented:
> 1. This is my fault, you just followed my suggestion from my previous comment about how to name the filter-creation helper functions (`metric-id-filter` etc.)... so, apologies.  But, when poking around with those functions in the IDE, I found myself hoping that the function names would all have a common prefix so that I could start typing `filter-` and then the editor would be able to show me a list of autocompletion choices, and from that I'd know exactly what my options were, as opposed to having to go find where those things live together in the code or look at docs, etc.  So, how would you feel about renaming those to something like `filter-metric-id` of `filter-by-metric-id` or `filter-with-metric-id` or something along those lines?

Any of those seem fine... `filter-with-metric-id`, `filter-with-url`, and `filter-with-url-and-method` might seem best to me. (naming--)

> 2. a) make the return value from that function have a top-level map with two keys in it, like `:metric-ids` and `:urls`, or some variant of that?  and then the individual metrics would live in nested maps underneath those?  

Yep, that definitely makes sense. Should we split out url-and-method timers too? Or just `:metric-ids` and `:urls`? If no `:metric-ids` have been specified in any requests, is it okay to include that key with `nil` under it?

> b) should we be putting the metric-id tuple into the results as another key/value pair in those maps?

I think putting the metric-id tuple into the results as another key/value pair makes a lot of sense. One concern I have though is for clojure, is it okay to normalize this to strings? I don't think that we have any way of knowing whether, when the metric-id was specified as a request option, each of the parts of the tuple was a string or a keyword.

Should we also do this (include the url and method) for the url timers? That seems like it would make sense, although it might also be more complicated because we can't just split on `.` (but I think still doable).

> c) should we be using the metric-id tuple as the key for the metrics that have a metric id?

I think I'm slightly against this/concerned about how this would work. Can you have a clojure array as the key of a map (that seems really weird then and not actually helpful)? Or would you have to turn it into a string?

If we do do this, I think that we should also do the same thing for the url metrics, for consistency. But I think that I'm pretty against it at this point.

>For b) - things I think we might want to do before release but am fine with not including in this PR; maybe they should be ticketed: 
1. Adding a filter fn for "all metrics with any metrics id" (this might be unnecessary if we reorganize the data structure returned from `get-client-metrics` so that there was just a nested map in there that had these metrics in it).

In the current version of the code, this can be done with `(metric-id-filter [])` - do you think it needs it's own separate function? I also have a commit (that I haven't pushed up to this PR because I think we said that work should be a separate ticket) to allow for doing `(metric-id-filter ["foo" "*"])` and getting all metric-id metrics that used a metric tuple that starts with `foo`.

> 2. Talk about what we want the final names to be for `with-*` and `bytes-read`.

Yeah, agreed we should do this... I'm terrible at naming. How do we come to a consensus on this or even figure out better names for things?

>3. Maybe play with graphite and validate that it's possible to access metrics that have funky characters in them, e.g. from a URL being embedded in a metric name

Makes sense to me. I've filed TK-353 for this.

>For c) - stuff that we should or could do but doesn't need to block the initial release per se:
1. Docs - and this doesn't have to be all @rlinehan , we already had a docs deficiency in this repo, it just seems like it's even more important now

I think I would argue for doing this before the initial release with all this metric stuff - docs for the metrics at least, not docs for everything else (although I agree that should be done too). I've filed for TK-354 for metrics docs, and TK-355 for general docs.

>2. Look into cleaning up schema for MetricFilter

How are you thinking this should be cleaned up? It could use a `schema/either` to get rid of the `contains?` crap (I hate `schema/either`, but what's currently there doesn't seem any better than that)... or were you thinking of something bigger?

>3. Ability to filter URL metrics via regex

I've filed TK-356 for this, and TK-357 for getting all the child metric-ids under a certain one.

I also had one outstanding question that I think I'd like to get answered before this is merged, I've just copied and pasted it from above (https://github.com/puppetlabs/clj-http-client/pull/51#discussion_r57980742) 

>One thing that is a little odd about having the "get"metrics APIs connected to a client object is that if you have one MetricRegistry in use by two client instances, calling getClientMetrics* on one of these clients will return metrics for requests from both clients. Looking at it from this point of view, I think this actually might be really confusing the way this is implemented and would be less confusing if the "get" metrics APIs operated on the MetricRegistry object instead.

cprice404 commented:
> filter-with-metric-id, filter-with-url, and filter-with-url-and-method might seem best to me. (naming--)

I'm good with those.

> Yep, that definitely makes sense. Should we split out url-and-method timers too? Or just :metric-ids and :urls?

Interesting - I think I like that idea of splitting out `url-and-method` stuff as well.

> If no :metric-ids have been specified in any requests, is it okay to include that key with nil under it?

I think we should include the key and then maybe put an empty map under it.  That way callers won't have to put any special-case stuff / `nil` checks, and it will work fine with any mapping / collection transformation functions that they want to use on it.

> I think putting the metric-id tuple into the results as another key/value pair makes a lot of sense. One concern I have though is for clojure, is it okay to normalize this to strings? I don't think that we have any way of knowing whether, when the metric-id was specified as a request option, each of the parts of the tuple was a string or a keyword.

Yeah, I think they'll have to be strings.

> Should we also do this (include the url and method) for the url timers? That seems like it would make sense, although it might also be more complicated because we can't just split on . (but I think still doable).

I like this idea, but I'd be willing to let that one slide out to a future release.  I would strongly suggest, though, that you think about doing all of this stuff by re-organizing the data structure that you are using to manage the timer objects.  Like, keep a nested map (or simple Java class) of timer objects that has the other data at the appropriate levels, and then dig into the appropriate places in the map to find the data you're looking for.  If you're doing any string splitting to find this data (metric ids *or* URLs), it's probably going to be fragile, and I'd treat that as a code smell.

If it is going to be too much work to re-organize the data structure that holds the timers right now, and you've thought through it enough to convince yourself that that could be done in a follow-up PR, we can ticket that and let it slide.

> Can you have a clojure array as the key of a map (that seems really weird then and not actually helpful)?

Yes, you can.  If it's not useful, though, then it's not an absolute requirement.  What I would say, though, is that whatever we use for the keys for these maps ideally would be something that a user might conceivably want to use to index into the map.  It's hard for me to imagine a use case where the full-on dropwizard metric name is a thing that they'd want to use for that, since they don't even really have any sane way to construct that piece of data other than iterating over the results of one of your functions... and in that case, they're probably more likely to just iterate over the values and ignore the keys, right?

> If we do do this, I think that we should also do the same thing for the url metrics, for consistency.

That makes sense, and TBH indexing into those maps by URL does seem like a more likely use case.

> In the current version of the code, this can be done with (metric-id-filter []) - do you think it needs it's own separate function?

Oh, neat!  I didn't realize that.  I don't think it needs to be its own separate function.  I might consider just supporting a no-arg signature of that fn, or accepting a `nil` there, instead of the empty vector.  Just trying to think of what might be the most intuitive.  But I don't have a super strong opinion and would be fine with keeping that part as-is if you think that's the most intuitive API to achieve that.

>  I also have a commit (that I haven't pushed up to this PR because I think we said that work should be a separate ticket) to allow for doing (metric-id-filter ["foo" "*"])

Cool.  Let's definitely push that out to a later date.  I'd like to discuss whether or not we want to use that syntax, vs. supporting some optional flag like `(metric-id-filter ["foo"] {:include-child-metrics true})` or something along those lines.  Whenever possible I prefer to avoid special-casing sentinel values and instead codify something that fits more concretely into the function signature or type system.

> I think I would argue for doing this before the initial release with all this metric stuff - docs for the metrics at least, not docs for everything else (although I agree that should be done too).

I don't have any problem with that unless there's someone chomping at the bit to use the feature and we haven't found time to write the docs.  I agree it's better, in an ideal world.

> How are you thinking this should be cleaned up? It could use a schema/either to get rid of the contains? crap (I hate schema/either ...

I think we could probably do a `schema/conditional`, and then just have the predicates be checks against the sub-schemas.  It's still gross but for some reason it strikes me as significantly less gross (or at least more thorough) than something with `contains?` logic.  This isn't really a huge deal though as long as we don't treat this schema as public API yet.

> One thing that is a little odd about having the "get"metrics APIs connected to a client object is that if you have one MetricRegistry in use by two client instances, calling getClientMetrics* on one of these clients will return metrics for requests from both clients. Looking at it from this point of view, I think this actually might be really confusing the way this is implemented and would be less confusing if the "get" metrics APIs operated on the MetricRegistry object instead.

Hm.  Yeah, I must have missed that comment before.  That does seem crappy.  I'd be OK with switching those functions to work on a MetricRegistry object instead, then.  I think my preference would still be to have the MetricRegistry passed in as part of the client constructor, though, and then have an accessor function that would let you grab it back out of the client object?  As opposed to either a) forcing the user to always manage their own reference to it, and/or b) having it be passed in on a per-request basis?

Thanks a bunch for filing all those jira tickets.  We should probably create an epic to group them at this point, since there are quite a few that we've stacked up.  WDYT?

rlinehan commented:
Closing in favor of #52.
